File: 0637_average_of_levels.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque

class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        Q = deque([deque([root])])
        result = []
        while Q:
            node = Q.popleft()
            if not node: break
            else: Q.append(deque())
            nums = [] 
            while node:
                obj = node.popleft()
                nums.append(obj.val)
                if obj.left:
                    Q[0].append(obj.left)
                if obj.right:
                    Q[0].append(obj.right)
            result.append(sum(nums)/len(nums))
        return result

# Alternative solution

class Solution1:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        storage = []
        depth = 0
        self.dfs(root, storage, depth)
        return [s/l for s, l in storage]
        
    def dfs(self, node, storage, depth):
        if node:
            if len(storage) <= depth:
                storage.append([0, 0])
            storage[depth][0] += node.val
            storage[depth][1] += 1
            self.dfs(node.left, storage, depth + 1)
            self.dfs(node.right, storage, depth + 1)

# Alternative solution

from collections import deque

class Solution2:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        Q = deque([root])
        result = []
        while Q:
            n = len(Q)
            summ = 0
            for _ in range(n):
                node = Q.popleft()
                summ += node.val
                if node.left:
                    Q.append(node.left)
                if node.right:
                    Q.append(node.right)
            result.append(summ/n)
        return result

File: 0501-1000/0637_averageOfLevels.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> averages = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>() {{add(root);}};

        while (!queue.isEmpty()) {
            int length = queue.size();
            double levelSum = 0;

            for (int i = 0; i < length; i++) {
                TreeNode node = queue.remove();
                levelSum += node.val;

                if (node.left != null) {
                    queue.add(node.left);
                }

                if (node.right != null) {
                    queue.add(node.right);
                }
            }

            averages.add(levelSum/length);
        }

        return averages;
    }
}
